[WIP] Refactor and expand network tests

dperny opened the pull request:
This PR encompasses about a week's worth of work
- Refactors common code out of network tests
- Adds network test for global services
- Adds network test for internal services using an external service as a proxy.

Relies on #35.

dongluochen commented on tests/network_test.go:
```diff
@@ -21,22 +24,102 @@ import (
+			func() {
+				// TODO(dperny) consider breaking out into separate function
+				// lock the mutex to synchronize access to the map
+				mu.Lock()
```
If `mu` is for the map only, you may want to just wrap it around access to the map.

dperny replied:
i think in this case i wanted to either poll or check but not both. i explicitly wanted to trade one goroutine for another, not just at map access time.

dongluochen commented on tests/network_test.go:
```diff
@@ -47,31 +130,101 @@ func TestNetworkExternalLb(t *testing.T) {
+	//create the service and make sure it comes up
+	service, err := cli.ServiceCreate(testContext, spec, types.ServiceCreateOptions{})
+	require.NoError(t, err, "Error creating service")
+	ctx, cancel := context.WithTimeout(testContext, 30*time.Second)
```
define a constant like
https://github.com/docker/docker/blob/master/integration-cli/docker_api_swarm_test.go#L22

dperny commented on tests/network_test.go:
```diff
@@ -21,22 +24,102 @@ import (
+			// if we don't sleep, we'll starve the check function. we stop
+			// just long enough for the system to schedule the check function
+			// TODO(dperny): figure out a cleaner way to do this.
+			time.Sleep(5 * time.Millisecond)
```
It seems like I should call `runtime.Gosched()` right before I call that sleep?

dperny commented:
@dongluochen 
- I refactored the network tests to put the tests at the top and the utilities at the bottom. 
- I added a test for talking to one service from every node.
- I added a prep state to pre-pull every needed image to the cluster.

dongluochen commented:
@dperny Sorry I missed your update. Please rebase.
